### ISSUE TITLE
Warnings when compiling libyang

### ISSUE THREAD

**Issue by igortern** [CESNET/libyang#1300]
Hi,
The following compiler warnings appear when building libyang-devel using gcc-9 compiler. Most of the warnings are false positives, but warnings in xpath.c appear to indicate real bugs.

Regards,
Igor

-- Build files have been written to: /home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/build
Scanning dependencies of target compat
[  1%] Building C object CMakeFiles/compat.dir/compat/compat.c.o
[  1%] Built target compat
Scanning dependencies of target yangobj
[  2%] Building C object CMakeFiles/yangobj.dir/src/common.c.o
[  3%] Building C object CMakeFiles/yangobj.dir/src/context.c.o
[  4%] Building C object CMakeFiles/yangobj.dir/src/log.c.o
[  6%] Building C object CMakeFiles/yangobj.dir/src/hash_table.c.o
[  7%] Building C object CMakeFiles/yangobj.dir/src/resolve.c.o
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/resolve.c: In function 'resolve_must':
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/resolve.c:6288:27: warning: 'must' may be used uninitialized in this function [-Wmaybe-uninitialized]
 6288 |         if (lyxp_eval(must[i].expr, node, LYXP_NODE_ELEM, lyd_node_module(node), &set, LYXP_MUST)) {
      |                           ^
In file included from /home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/resolve.c:25:
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/resolve.c: In function 'resolve_identref':
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/build/src/common.h:220:5: warning: 'der' may be used uninitialized in this function [-Wmaybe-uninitialized]
  220 |     ly_vlog(ctx, code, elem_type, elem, ##args);
      |     ^~~~~~~
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/resolve.c:5846:23: note: 'der' was declared here
 5846 |     struct lys_ident *der, *cur;
      |                       ^~~
[  8%] Building C object CMakeFiles/yangobj.dir/src/validation.c.o
[  9%] Building C object CMakeFiles/yangobj.dir/src/xml.c.o
[ 11%] Building C object CMakeFiles/yangobj.dir/src/parser.c.o
[ 12%] Building C object CMakeFiles/yangobj.dir/src/parser_yin.c.o
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/parser_yin.c: In function 'fill_yin_unique':
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/parser_yin.c:2048:19: warning: 'c' may be used uninitialized in this function [-Wmaybe-uninitialized]
 2048 |             *vaux = c;
      |             ~~~~~~^~~
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/parser_yin.c: In function 'fill_yin_deviation':
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/parser_yin.c:2982:35: warning: 'k' may be used uninitialized in this function [-Wmaybe-uninitialized]
 2982 |                         } else if (dev_target->ext[j]->insubstmt_index > k) {
      |                                   ^
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/parser_yin.c: In function 'fill_yin_type':
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/parser_yin.c:1120:24: warning: 'v_' may be used uninitialized in this function [-Wmaybe-uninitialized]
 1120 |                     if (v_ != enms_sc[j].value) {
      |                        ^
[ 13%] Building C object CMakeFiles/yangobj.dir/src/parser_xml.c.o
[ 14%] Building C object CMakeFiles/yangobj.dir/src/parser_json.c.o
[ 16%] Building C object CMakeFiles/yangobj.dir/src/parser_lyb.c.o
[ 17%] Building C object CMakeFiles/yangobj.dir/src/parser_yang_bis.c.o
[ 18%] Building C object CMakeFiles/yangobj.dir/src/parser_yang_lex.c.o
[ 19%] Building C object CMakeFiles/yangobj.dir/src/parser_yang.c.o
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/parser_yang.c: In function 'yang_fill_unique':
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/parser_yang.c:540:19: warning: 'c' may be used uninitialized in this function [-Wmaybe-uninitialized]
  540 |             *vaux = c;
      |             ~~~~~~^~~
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/parser_yang.c: In function 'yang_deviate_delete_unique':
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/parser_yang.c:2048:19: warning: 'k' may be used uninitialized in this function [-Wmaybe-uninitialized]
 2048 |         } else if (list->ext[j]->insubstmt_index > k) {
      |                   ^
[ 20%] Building C object CMakeFiles/yangobj.dir/src/tree_schema.c.o
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/tree_schema.c: In function 'lys_features_change':
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/tree_schema.c:3976:46: warning: 'failj' may be used uninitialized in this function [-Wmaybe-uninitialized]
 3976 |                faili == -1 ? module->features[failj].name : module->inc[faili].submodule->features[failj].name, failk);
      |                                              ^
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/tree_schema.c:3976:72: warning: 'faili' may be used uninitialized in this function [-Wmaybe-uninitialized]
 3976 |                faili == -1 ? module->features[failj].name : module->inc[faili].submodule->features[failj].name, failk);
      |                                                                        ^
[ 22%] Building C object CMakeFiles/yangobj.dir/src/tree_data.c.o
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/tree_data.c: In function 'lyd_insert_nextto':
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/tree_data.c:4720:33: warning: 'last' may be used uninitialized in this function [-Wmaybe-uninitialized]
 4720 |             sibling->next->prev = last;
      |             ~~~~~~~~~~~~~~~~~~~~^~~~~~
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/tree_data.c: In function 'lyd_new_path':
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/tree_data.c:2018:53: warning: 'yang_data_name_len' may be used uninitialized in this function [-Wmaybe-uninitialized]
 2018 |                 str = strndup(path, (yang_data_name + yang_data_name_len) - path);
      |                                                     ^
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/tree_data.c: In function 'lyd_defaults_add_unres':
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/tree_data.c:8187:23: warning: 'data_tree_parent' may be used uninitialized in this function [-Wmaybe-uninitialized]
 8187 |                 set = lyd_find_instance(data_tree_parent ? data_tree_parent : data_tree_sibling, msg_op);
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[ 23%] Building C object CMakeFiles/yangobj.dir/src/plugins.c.o
[ 24%] Building C object CMakeFiles/yangobj.dir/src/printer.c.o
[ 25%] Building C object CMakeFiles/yangobj.dir/src/xpath.c.o
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/xpath.c: In function 'get_node_pos':
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/xpath.c:1321:8: warning: 'elem' may be used uninitialized in this function [-Wmaybe-uninitialized]
 1321 |     if (!elem) {
      |        ^
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/xpath.c: In function 'cast_string_recursive':
/home/igort/Perforce/OLT/aspen/main/build/host-sim/third_party/libyang/libyang-devel/src/xpath.c:473:9: warning: 'ptr' may be used uninitialized in this function [-Wmaybe-uninitialized]
  473 |         } while ((line = strtok_r(NULL, "\n", &ptr)));
      |         ^

**Comment by michalvasko** [comment]
Yeah, sorry about that. They are all false positives caused by the usage of `-Og`, they disappear if you remove it. But it is needed for ABI check target.

**Comment by igortern** [comment]
Yeah, i didn't check all of them, but warnings in xpath.c appear to be real bugs

**Comment by michalvasko** [comment]
Like I said, they are all false positives. But for the peace of mind, I have just fixed them.
